### PR TITLE
gh-77578: IDLE Help - let users control font size

### PR DESCRIPTION
**Pull request by csabella**
- Get initial font size from configuration.
- Allow for dynamically changing font size via keyboard or mousewheel.

* Issue: gh-77578

### CODE REVIEW

File: Lib/idlelib/idle_test/test_help.py [hunk 1]
```diff
@@ -30,5 +40,75 @@ def test_line1(self):
         text = self.frame.text
         self.assertEqual(text.get('1.0', '1.end'), ' IDLE ')
 
+
+class HelpTestTest(unittest.TestCase):
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_help.py]: typo - HelpTextTest

File: Lib/idlelib/help.py [hunk 2]
```diff
@@ -160,26 +160,33 @@ def __init__(self, parent, filename):
...
         self.parser.feed(contents)
+
+        self.bind_events()
         self['state'] = 'disabled'
+        self.focus_set()
+
+    def create_fonts(self):
+        "Create fonts to be used with tags."
+        self.base_size = idleConf.GetOption('main', 'EditorWindow',
+                                            'font-size', type='int')
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: This sets base_size to 10 unless user overrides.  Test must not depend on any particular value.

File: Lib/idlelib/help.py [hunk 3]
```diff
@@ -191,6 +198,52 @@ def findfont(self, names):
...
+
+    def bind_events(self):
+        "Bind events."
+        # Zoom out works with or without shift.
+        self.event_add('<<zoom-text-out>>',
+                       *['<Control-Key-equal>', '<Control-Key-plus>'])
+        self.bind('<<zoom-text-out>>', lambda e: self.zoom(e, 'out'))
+
+        self.event_add('<<zoom-text-in>>', '<Control-Key-minus>')
+        self.bind('<<zoom-text-in>>', lambda e: self.zoom(e, 'in'))
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: I don't think we need the indirection of a pseudoevent or the lambda.
self.bind('<Control-Key-Equal>', self.zoom)
etc should work as zoom can switch on event.type (keypress, button, or wheel).

File: Lib/idlelib/help.py [hunk 4]
```diff
@@ -191,6 +198,52 @@ def findfont(self, names):
...
+
+        # Windows and Mac use MouseWheel.
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: But they use opposite signs and magnitudes.  From turtledemo.__main__
        # For wheel up, event.delta = 120 on Windows, -1 on darwin.
I want to use the debugged code from that file (which I helped write ;-) as a starting point, but it can be simplified to one event handler (zoom) and one new size handler (here scale fontsize).

I also want to consider that the basic handling should be the same for editor windows. Perhaps we can factor out a FontSizer class, with the event handlers, that is initialized with the widget to bind methods to and a method to call with the new font size.  This would be easier with standardized attribute names.

File: Lib/idlelib/help.py [hunk 5]
```diff
@@ -191,6 +198,52 @@ def findfont(self, names):
...
+
+    def scale_fontsize(self, base=12):
+        "Scale tag sizes based on the size of normal text."
+        # Scale percentages are from Sphinx classic.css.
+        scale = {'normal': 1.0, 'h6': 1.0, 'h5': 1.1, 'h4': 1.2, 'h3': 1.4,
+                 'h2': 1.6, 'h1': 2.0, 'em': 1.0, 'pre': 1.0, 'preblock': 0.9}
+        for tag in self.fonts:
+            self.fonts[tag]['size'] = int(base * scale[tag])
+
+    def zoom(self, event, how='out'):
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: Delete 'how' and switch on event.type.

File: Lib/idlelib/help.py [hunk 6]
```diff
@@ -191,6 +198,52 @@ def findfont(self, names):
...
+        self.bind('<<zoom-text-in>>', lambda e: self.zoom(e, 'in'))
+
+        # Windows and Mac use MouseWheel.
+        self.bind('<Control-MouseWheel>', lambda e: self.zoom(e, 'wheel'))
+        # Linux uses Button 4 (scroll down) and Button 5 (scroll up).
+        self.bind('<Control-Button-4>', lambda e: self.zoom(e, 'wheel'))
+        self.bind('<Control-Button-5>', lambda e: self.zoom(e, 'wheel'))
+
+    def scale_fontsize(self, base=12):
+        "Scale tag sizes based on the size of normal text."
```
Comment (terryjreedy) [inline Lib/idlelib/help.py:233]: No default.  Set self.base_size to base passed in.

File: Lib/idlelib/help.py [hunk 7]
```diff
@@ -191,6 +198,52 @@ def findfont(self, names):
...
+        # Scale percentages are from Sphinx classic.css.
+        scale = {'normal': 1.0, 'h6': 1.0, 'h5': 1.1, 'h4': 1.2, 'h3': 1.4,
+                 'h2': 1.6, 'h1': 2.0, 'em': 1.0, 'pre': 1.0, 'preblock': 0.9}
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: h6 to h4 are not needed.  'h1': 20 seems too big as current 20/12 = 5/3 = 1.6 seems sufficient.  ditto for h2 18/12 = 3/2 = 1.5, h3 = 15/12 = 5/4 = 1.2.  The current ratios result in text the nearly matches the Firefox version.

File: Lib/idlelib/help.py [hunk 8]
```diff
@@ -191,6 +198,52 @@ def findfont(self, names):
...
+        for tag in self.fonts:
+            self.fonts[tag]['size'] = int(base * scale[tag])
+
+    def zoom(self, event, how='out'):
+        "Handle zooming text in/out."
+        if how == 'wheel':
+            # Button 5 or negative delta is when mouse wheel is scrolled down.
+            how = 'in' if event.num == 5 or event.delta < 0 else 'out'
+        factor = 1.1 if how == 'out' else 0.9
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: I would rather just increase/decrease the integer size by 1, but within bounds.
```
self.scale_fontsize(max(self.base_size - 1, MINIMUM_FONT_SIZE) if <decrease> else
                              min(self.base_size - 1, MAXIMUM_FONT_SIZE))

File: Lib/idlelib/idle_test/test_help.py [hunk 9]
```diff
@@ -30,5 +40,75 @@ def test_line1(self):
...
+        tests = (('<<zoom-text-out>>', 0, zoomout),
+                 ('<<zoom-text-in>>', 0, base),
+                 ('<<zoom-text-in>>', 0, zoomin),
+                 ('<Control-MouseWheel>', 0, base),
+                 ('<Control-MouseWheel>', -120, zoomin),
+                 ('<Control-MouseWheel>', 120, base),
+                 ('<Control-Button-4>', 0, zoomout),
+                 ('<Control-Button-5>', 0, base))
+
+        # Base size starts at 12.
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_help.py]: You set it to the default editor text size, which is 10 on buildbots.

File: Lib/idlelib/idle_test/test_help.py [hunk 10]
```diff
@@ -30,5 +40,75 @@ def test_line1(self):
...
+        sizes = [text.fonts[tag]['size'] for tag in tags]
+        eq(text.base_size, base[0])
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_help.py]: 10 is not 12.

File: Lib/idlelib/help.py [hunk 11]
```diff
@@ -160,26 +160,33 @@ def __init__(self, parent, filename):
...
+        self.focus_set()
+
+    def create_fonts(self):
+        "Create fonts to be used with tags."
+        self.base_size = idleConf.GetOption('main', 'EditorWindow',
+                                            'font-size', type='int')
+        normalfont = self.findfont(['TkDefaultFont', 'arial', 'helvetica'])
+        fixedfont = self.findfont(['TkFixedFont', 'monaco', 'courier'])
+
+        self.fonts = fonts = {}
```
Comment (terryjreedy) [inline Lib/idlelib/help.py:202]: I don't really like the doubling of the configuration code, but I have not thought of anything better yet.

**Review by terryjreedy — CHANGES_REQUESTED** [review]

File: Lib/idlelib/idle_test/test_help.py (hunk 1 above)
Comment (csabella) [reply Lib/idlelib/idle_test/test_help.py]: Done

File: Lib/idlelib/help.py (hunk 2 above)
Comment (csabella) [reply Lib/idlelib/help.py]: Thanks.  I've changed the test to set this value.

File: Lib/idlelib/help.py (hunk 3 above)
Comment (csabella) [reply Lib/idlelib/help.py]: I originally did this because I wasn't sure about hardcoding the checks for the 'plus' and 'minus' keys, if those would be configurable at some point.  I've changed it to look for the keys and I suppose it can be changed later to use the keys attached to the binding.

File: Lib/idlelib/help.py (hunk 4 above)
Comment (csabella) [reply Lib/idlelib/help.py]: I hadn't seen the turtledemo code before I made my PR, but it was helpful to look at it now.

Created the FontSizer class.  I tried to do it in a way as to not rely on the names that the widget uses.  The only thing that is required is that the font used by the widget be an actual `Font()` and not just defined on the fly.

File: Lib/idlelib/help.py (hunk 5 above)
Comment (csabella) [reply Lib/idlelib/help.py]: Done.  I like it much better this way.  :-)

File: Lib/idlelib/help.py (hunk 6 above)
Comment (csabella) [reply Lib/idlelib/help.py:233]: With the refactoring of the class for FontSizer, I removed self.base_size.

File: Lib/idlelib/help.py (hunk 7 above)
Comment (csabella) [reply Lib/idlelib/help.py]: I had originally added the extra values just to show the progression, even though they weren't needed.  Now I've removed them and updated the ratios.

File: Lib/idlelib/help.py (hunk 8 above)
Comment (csabella) [reply Lib/idlelib/help.py]: Done

File: Lib/idlelib/idle_test/test_help.py (hunk 9 above)
Comment (csabella) [reply Lib/idlelib/idle_test/test_help.py]: Fixed to set this value during the test.

File: Lib/idlelib/idle_test/test_help.py (hunk 10 above)
Comment (csabella) [reply Lib/idlelib/idle_test/test_help.py]: Done

File: Lib/idlelib/help.py (hunk 11 above)
Comment (csabella) [reply Lib/idlelib/help.py:202]: I didn't make any changes for this.

File: Lib/idlelib/idle_test/test_help.py (hunk 1 above)
Comment (terryjreedy) [reply Lib/idlelib/idle_test/test_help.py]: check

File: Lib/idlelib/help.py (hunk 2 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: Set to 12 *in the test*

File: Lib/idlelib/help.py [hunk 12]
```diff
@@ -150,6 +157,58 @@ def handle_data(self, data):
...
+        self.callback = callback
+        self.bind_events()
+
+    def bind_events(self):
+        "Bind events to the widget."
+        shortcut = 'Command' if darwin else 'Control'
+        # Zoom out works with or without shift.
+        self.widget.event_add(
+                '<<zoom-text-out>>',
+                *[f'<{shortcut}-Key-equal>', f'<{shortcut}-Key-plus>'])
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: f(*[1,2]) is exactly the same as f(1,2).  Using * in a call is for when one has a name, not a display.
so pass `f'<{shortcut}-Key-equal>', f'<{shortcut}-Key-plus>'`.
I would still like to remove the pseudoevents.

File: Lib/idlelib/help.py (hunk 3 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: I think, if anything, we have too many key configurations.  zoom has been changed.  See new review comment.

File: Lib/idlelib/help.py [hunk 13]
```diff
@@ -150,6 +157,58 @@ def handle_data(self, data):
             self.text.insert('end', d, (self.tags, self.chartags))
 
 
+class FontSizer:
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: (Note to me: Review this more after manual test.)

File: Lib/idlelib/help.py (hunk 4 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: Still need to review.

File: Lib/idlelib/help.py (hunk 5 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: check

File: Lib/idlelib/help.py (hunk 8 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: check

**Review by terryjreedy — CHANGES_REQUESTED** [review]
You can leave the one suggested code change until I retest and recheck code.  3.7.0c1 is scheduled next Tues, but I don't want to rush this without test on Mac.  I really must try out devguide instructions with a small MacBook portable I was given.

File: Lib/idlelib/help.py (hunk 13 above)
Comment (taleinat) [inline Lib/idlelib/help.py]: The separation between the zoom handling and the event binding should be better.

I suggest creating three event handlers: `zoom_in`, `zoom_out`, and `zoom_mousewheel`. They would all internally call a common `zoom()` method, which wouldn't be an event handler, and would receive just a direction parameter.

**Review by taleinat — DISMISSED** [review]

File: Lib/idlelib/help.py (hunk 13 above)
Comment (csabella) [reply Lib/idlelib/help.py]: @taleinat, thanks for the suggestions!  I've made the changes and added tests for the new functions.

File: Lib/idlelib/help.py (hunk 12 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: The unneeded *[..] is still present here and on line 185.

turtledemo.__main__ does not have these pseudoevents.  The only reason I can think of now to add them is if they were to make testing significantly easier and more dependable.

File: Lib/idlelib/help.py [hunk 14]
```diff
@@ -176,31 +176,39 @@ def bind_events(self):
...
+        "Make font size larger."
+        self.zoom(increase=True)
+
+    def zoom_out(self, event):
+        "Make font size smaller."
+        self.zoom(increase=False)
+
+    def zoom_mousewheel(self, event):
+        "Adjust font size based on mouse wheel direction."
+        wheel_up = {EventType.MouseWheel: event.delta > 0,
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: This leaves out '== darwin', which I believe is correct for font resizing even if not correct for scrolling.

File: Lib/idlelib/help.py [hunk 15]
```diff
@@ -176,31 +176,39 @@ def bind_events(self):
         shortcut = 'Command' if darwin else 'Control'
         # Zoom out works with or without shift.
         self.widget.event_add(
-                '<<zoom-text-out>>',
+                '<<zoom-text-in>>',
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: I guess the pseudoevent error was masked since the following line has the same error and the direction was recalculated (in former version of zoom).

**Review by terryjreedy — CHANGES_REQUESTED** [review]
We started with this code, extracted from turtledemo.__main__.
```
# Menu bindings (which we might use for context menu)
        menu.add_command(label="Decrease (C-'-')", command=self.decrease_size,
                         font=menufont)
        menu.add_command(label="Increase (C-'+')", command=self.increase_size,
                         font=menufont)
# Key/mouse bindings
        text.bind_all('<%s-minus>' % shortcut, self.decrease_size)
        text.bind_all('<%s-underscore>' % shortcut, self.decrease_size)
        text.bind_all('<%s-equal>' % shortcut, self.increase_size)
        text.bind_all('<%s-plus>' % shortcut, self.increase_size)
        text.bind('<Control-MouseWheel>', self.update_mousewheel)
        text.bind('<Control-Button-4>', self.increase_size)
        text.bind('<Control-Button-5>', self.decrease_size)
# Event/action functions
darwin = sys.platform == 'darwin'

    def set_txtsize(self, size):
        txtfont[1] = size
        self.text['font'] = tuple(txtfont)
        self.output_lbl['text'] = 'Font size %d' % size

    def decrease_size(self, dummy=None):
        self.set_txtsize(max(txtfont[1] - 1, MINIMUM_FONT_SIZE))
        return 'break'

    def increase_size(self, dummy=None):
        self.set_txtsize(min(txtfont[1] + 1, MAXIMUM_FONT_SIZE))
        return 'break'

    def update_mousewheel(self, event):
        # For wheel up, event.delta = 120 on Windows, -1 on darwin.
        # X-11 sends Control-Button-4 event instead.
        if (event.delta < 0) == (not darwin):
            return self.decrease_size()
        else:
            return self.increase_size()
```
As far as I know, menu, keyboard, and mouse methods work on all three major systems.  There are no pseudoevents.  To the extent possible, the meaning of actions is determined by their bindings.

Note: the code above is extracted from 4 different locations.  This is the first time I have looked at it altogether.

At my request, Cheryl combined and consolidated the functions into one zoom handler.  At Tal's request, she partly undid what I had requested.  Zoom() has been separated into multiple functions again.  However, I prefer the original code to the result.  If we don't agree on reverting to the original design, lets discuss.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
github threw away or hid my my review summary.  Will try to find or recreate.

Now it is visible again.

File: Lib/idlelib/help.py (hunk 12 above)
Comment (csabella) [reply Lib/idlelib/help.py]: I tried back in May and I've been trying for a few days to get the tests to work without the virtual events.  Even though it seems like the text widget has focus, doing a `text.event_generate('<Control-Key-plus>')` just won't activate the binding.  I'm sure I'm just missing something obvious, but I cannot figure out what it is.  :-(

File: Lib/idlelib/help.py (hunk 15 above)
Comment (csabella) [reply Lib/idlelib/help.py]: While refamiliarizing myself with this code, I kept getting confused on what 'zoom out' and 'zoom in' meant and added a comment to the new methods in help.py to say which means larger text and which means smaller text.  Once I did that, I realized I had it backwards in the definitions.

File: Lib/idlelib/help.py (hunk 14 above)
Comment (csabella) [reply Lib/idlelib/help.py]: I wasn't sure when that logic needed to be applied and when it didn't.  So, it's only adjusted on scrolling and not on other uses of the mousewheel?

**Comment by taleinat** [comment]
My apologies, @terryjreedy and @csabella. I wasn't aware of Terry's request to consolidate the zoom handling into a single handler. Cheryl, please do as Terry has requested, as he has the final say on these matters.

File: Lib/idlelib/help.py (hunk 14 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: I tested turtledemo a month ago and font resizing works the same as on Windows.  Moving the top of the wheeI up increases size.  I don't actually know if this is the correct convention on Mac, but in the absence of information otherwise, including the absence of complaint about turtledemo, I will consider it correct.  I may have tested command-wheel-up in Safari and will recheck tomorrow.

When we copied to logic for scrolling, the result was buggy since the convention on Mac, which I know I verified with Safari, is the opposite of Windows (and Linux, I believe).  I will try to find and point you to the tracker discussion.

File: Lib/idlelib/help.py (hunk 15 above)
Comment (terryjreedy) [reply Lib/idlelib/help.py]: I agree that these names are potentially confusing and propose above (May 17 comment on line 180) 'make_font_bigger' and 'make_font_smaller'.

**Comment by taleinat** [comment]
@terryjreedy, on the contrary, I should just take more care to read all of the previous discussion. You've got more than enough on your plate, you shouldn't bother to write summaries when the reviews are there to be read.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Current requested changes: Reverse previous request to combine turtledemo functions into zoom().  However, add psuedoevents '<<increase_font_size>>' and '<<decrease_font_size>>', use the same names for the handler functions.  Don't use *[] in event add.  Use original mouse logic.

Use pseudoevents for testing.  Don't worry about key-mouse events now.

**Comment by terryjreedy** [comment]
The above is what I meant by a concise summary.  It should help Cheryl to rewrite and me to review when she does..

**Comment by csabella** [comment]
@terryjreedy, thanks for summarizing the requested changes.  It did help make it easier to not miss something (and hopefully I did all the requests the way you wanted).

One thing that I've been trying to figure out is whether or not I have a memory leak with the Font objects that have been created in HelpText.  If you add a print statement, these Font objects are recreated every time that `IDLE Help` is clicked on.  I'm wondering if it would be better to explicitly name them and then find the existing objects instead of creating new ones every time.

File: Lib/idlelib/help.py [hunk 16]
```diff
@@ -2,7 +2,7 @@
 Contents are subject to revision at any time, without notice.
 
 
-Help => About IDLE: diplay About Idle dialog
+Help => About IDLE: display About Idle dialog
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: I guess spell checkers do not catch everything :-(,

**Comment by terryjreedy** [comment]
I intend to review again tomorrow (Sunday) and will consider your question.  

My 4-line patch in #9502 conflicts with this one.  I believe merging this first would be much easier, as it would be trivial and perhaps easiest for me to make a new PR on top of this.

**Comment by terryjreedy** [comment]
On Windows: Control- minus or plus on main keyboard and numpad (independent of Numlock setting) and Control-wheel-down or up decrease or increase font size, as intended.  Tal, it would be nice if you could recheck this on Mac.

Initially, the window contracts or expands with font size, unless it is maximized.  If either of width or height is explicitly set with the mouse , that dimension is no longer changed with font size changes.  I don't remember if this is how tk behaves or if the new code imposes this.

If the window width is fixed and the font size increases enough to make a header line wrap, then using the TOC to go to the header puts the wrapped part at the top of the window.  For instance, go to
4.1 Additional help
sources
when 'sources' is wrapped displays 'sources' and cuts off the first part of the header.  Currently, the toc entries point to the end of the header, so I think pointing to the beginning should fix this.

File: Lib/idlelib/help.py [hunk 17]
```diff
@@ -150,6 +157,69 @@ def handle_data(self, data):
             self.text.insert('end', d, (self.tags, self.chartags))
 
 
+class FontSizer:
+    "Support dynamic widget font resizing."
+    def __init__(self, widget, callback=None):
+        """"Add font resizing functionality to widget.
+
+        Args:
+            widget: Tk widget with font attribute to size.
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: I strongly prefer 'text' to 'widget'.  I am dubious that we will ever use this for anything other than a Text widget, and if we do, 'text' would not hurt for 'widget with font attribute'.

File: Lib/idlelib/help.py [hunk 18]
```diff
@@ -39,6 +41,11 @@
 
 ## IDLE Help ##
 
+darwin = sys.platform == 'darwin'
+MINIMUM_FONT_SIZE = 6
+MAXIMUM_FONT_SIZE = 100
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: Move to textview.py, along with FontSizer.  See later comment on FontSizer.

File: Lib/idlelib/help.py [hunk 19]
```diff
@@ -150,6 +157,69 @@ def handle_data(self, data):
...
 
 
+class FontSizer:
+    "Support dynamic widget font resizing."
+    def __init__(self, widget, callback=None):
+        """"Add font resizing functionality to widget.
+
+        Args:
+            widget: Tk widget with font attribute to size.
+            callback: Function to call for additional font resizing
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: This hook is not needed for text widgets with just a font attribute, like the turtledemo, text viewer, and editor subclasses.  I now think we should have started with adding font sizing to the text viewer.  So lets move this class to textview.py and add FontSizer(self(.text)) to the appropriate place, and see if it works without a callback (preferred) or if a callback is required.  Also import FontSizer into help.py.

File: Lib/idlelib/help.py [hunk 20]
```diff
@@ -150,6 +157,69 @@ def handle_data(self, data):
...
+        self.widget.event_add(
+                '<<decrease_font_size>>',
+                f'<{shortcut}-Key-minus>', f'<{shortcut}-Key-underscore>')
+        self.widget.bind('<<decrease_font_size>>', self.decrease_font_size)
+
+        # Windows and Mac use MouseWheel.
+        self.widget.bind('<Control-MouseWheel>', self.update_mousewheel)
+        # Linux uses Button 4 (scroll down) and Button 5 (scroll up).
+        self.widget.bind('<Control-Button-4>', self.update_mousewheel)
+        self.widget.bind('<Control-Button-5>', self.update_mousewheel)
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: Does this really work on Linux?  IE, do these mouse-button mouse wheel events have deltas?  Even if so, I would prefer binding each concreate event to the correct abstract event in the event-add calls above.  turtledemo.__main__ has 
        text.bind('<Control-Button-4>', self.increase_size)
        text.bind('<Control-Button-5>', self.decrease_size)
This is the opposite of how I interpret your comment.  What we care about is consistent (top of) wheel up/down == size up/down, regardless of what how wheel up and down scroll.  The latter is the opposite on Windows and Mac; I don't know about Unix or even if consistent across all varieties.

File: Lib/idlelib/help.py [hunk 21]
```diff
@@ -150,6 +157,69 @@ def handle_data(self, data):
...
+    "Support dynamic widget font resizing."
+    def __init__(self, widget, callback=None):
+        """"Add font resizing functionality to widget.
+
+        Args:
+            widget: Tk widget with font attribute to size.
+            callback: Function to call for additional font resizing
+                based on widget's font attribute.
+        """
+        self.widget = widget
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: A basic problem with manipulating just font sizes is that `text['font'] = font_tuple` does not assign the font tuple to the font attribute, but is translated to test.configure(font=font_tuple), which in turn converts the Python tuple into a tk font.  So `text['font;]` does not return the tuple but the tk string id of the font, such as 'font17'.

turtledemo deals with this by initializing txtfont to txtfont to ['Lucida Console', 10, 'normal']  and text['font'] to tuple(txtfont).  (I am not sure if tuplelizing is required.)  The font change mechanism changes txtfont[1] and then text['font'].  I considered suggesting adding something similar here, but at least for editors, we want to accommodate font family changes.

File: Lib/idlelib/help.py [hunk 22]
```diff
@@ -150,6 +157,69 @@ def handle_data(self, data):
...
+
+        # Windows and Mac use MouseWheel.
+        self.widget.bind('<Control-MouseWheel>', self.update_mousewheel)
+        # Linux uses Button 4 (scroll down) and Button 5 (scroll up).
+        self.widget.bind('<Control-Button-4>', self.update_mousewheel)
+        self.widget.bind('<Control-Button-5>', self.update_mousewheel)
+
+    def set_text_size(self, new_size):
+        "Set the font size for this widget."
+        font = tkfont.Font(self.widget, name=self.widget['font'], exists=True)
```
Comment (terryjreedy) [inline Lib/idlelib/help.py]: I need to review and experiment with this and the other font functions before I finish reviewing this and the next two functions.

File: Lib/idlelib/help.py [hunk 23]
```diff
@@ -191,6 +272,24 @@ def findfont(self, names):
...
                 return name
 
+    def configure_tags(self):
+        "Configure tags used in parsing."
+        for tag in ('em', 'h1', 'h2', 'h3', 'pre', 'preblock'):
+            self.tag_configure(tag, font=self.fonts[tag])
+        self.tag_configure('pre', background='#f6f6ff')
+        self.tag_configure('preblock', lmargin1=25, borderwidth=1,
+                           relief='solid', background='#eeffcc')
+        for level in range(1, 5):
```
Comment (terryjreedy) [inline Lib/idlelib/help.py:229]: This is one more level than before.  I don't know if we use more than l1.  I may add some prints to the parser to fine out.  We do use everything else.

**Review by terryjreedy — COMMENTED** [review]
The difficulty I have had it that this PR addresses two different (sub)issues.  1. Create a font size class that can be used with any text widget.  2. Apply it to help viewer, which requires multiple size adjustments.

I would like to put the new class and its tests where they belongs, in textview and test_textview, and apply it there also.  In the meanwhile, I should study the font and tag_configure functions.

In other words, approved in principle, but with further changes.

File: Lib/idlelib/help.py (hunk 19 above)
Comment (csabella) [reply Lib/idlelib/help.py]: It was designed to work without a callback for a just a text widget with a font and I had tried to write the `FontSizer` tests in a way that showed that (that is, the test didn't define a callback).  

The callback was written in order to allow for scaling of the other size fonts defined in `help.py`.  The `test_scale_tagfonts` and `test_resizing_callback` were meant to show that.  Sorry it wasn't clear.

Here was the issue:
In `help.py`, there are several different tags (h3, h2, h1) that have, by definition, different sizes.  So, I needed a way to change all the font sizes when the event was triggered.  With FontSizer being extracted as its own class, that class would only operate on the font defined within the text widget and not other ancillary fonts somehow related to that widget (in tags).  
Options:
1. What I really wanted to do was to have the reassignment of the font size in FontSizer be the trigger event.  That way when the font size changed, the HelpText class could watch for that event and call `scale_tagfonts`.  Even though there is a `Configure` event available, it seemed more to be for window resizing rather than font changes.
2.  A second option I considered was to add an event timer like in CodeContext, but I thought this wouldn't be smooth since one font would change immediately, but the others would be on a (possible) delay.  Plus, you didn't really like the timers in CodeContext.
3.  Add a callback that can be sent to FontSizer for any widgets that needed to do their own processing after the main font was changed.  This was what I implemented.
4.  Thinking about it now, I've come up with another idea.  Maybe there's a way for FontSizer to look at all the tags defined to `text` and modify the fonts for those tags.  I'll have to see if that's possible.

Did you have other ideas in mind?

File: Lib/idlelib/help.py (hunk 23 above)
Comment (csabella) [reply Lib/idlelib/help.py:229]: Not sure what you mean by 'one more level' than before?  Before, it did levels l1, l2, l3, and l4.  Debugging shows that down to l3 is used.

**Comment by csabella** [comment]
@terryjreedy  Hi Terry, I'm not sure what the blocker on this one is.  Should I split the font sizer in the textview into its own PR and then just handle the help resizing here?

**Comment by terryjreedy** [comment]
Me getting back to this.  I spent November before the holiday catching up on needed doc changes.  The last week I have been testing and discussing squeezer with Tal and looking at tk/tkinter/IDLE issues on Mac.  This is next.

**Comment by terryjreedy** [comment]
If I understand correctly,  742ab72 is the new patch since my review, and the force-push created squashed commit  8983368.  742ab72 moved FontSizer and tests 'as-is' to textview.

**Comment by terryjreedy** [comment]
Separate PRs for Fontsizer in textview and use of Fontsizer in help would have made sense in the beginning.  However, my live test (on Windows with 8.6.9, upgraded yesterday) shows that that it works* in both (I used About IDLE News for text view).  So no point in pulling them apart now.  I can, however, review the changes separately more easily than before.

* The only glitch is that making the font big enough caused the [Close] box to disappear.  It is not really needed, but it should not come and go.  I hope/presume that some pack/grid setting should prevent this.  Feel free to experiment.

When we use this in the editor, the status bar must not disappear.  bpo-34864 is about the status bar disappearing on Mac when a system setting is changed.  I believe we determined that this was not IDLE specific, but it would be nice if changing geometry settings could prevent that too.

**Comment by csabella** [comment]
> If I understand correctly, 742ab72 is the new patch since my review, and the force-push created squashed commit 8983368. 742ab72 moved FontSizer and tests 'as-is' to textview.

I wasn't really sure what happened with that force-push.  I had a conflict on help.py and that's what happened after following the steps from the devguide, but I don't remember it doing that before.  Maybe it's new workflow.  If I click on 8983368 above, then that shows the moving of FontSizer and tests to textview, so I think that was the latest changes since the last review.  742ab72 was just a quick commit because I messed up the directions of increase/decrease (again).

**Comment by csabella** [comment]
I pushed a commit for the issue with losing the 'Close' button.  The documentation said that it will pack the widgets in the given packing order and if a previous widget took up all the space, then other widgets may not display.  I tried a few things that didn't work, but packing the button first and then the frame seemed to fix it.  It seemed to be the simplest fix.

**Comment by vstinner** [comment]
I removed the " needs backport to 3.6" label, the 3.6 branch no longer accept bugfixes (only security fixes are accepted): https://devguide.python.org/#status-of-python-branches

**Comment by csabella** [comment]
I pushed a commit to allow for resizing fonts whether they are defined as a Font object or defined as a tuple.  I also add one line to `editor.py` to show it working.  :-)  I know that wouldn't be part of this PR, but I wanted to show that it works.

**Comment by csabella** [comment]
I have removed the unrelated changes to `editor.py` and also removed the `DO NOT MERGE` label since it was added because of that change.

**Comment by taleinat** [comment]
Ping, @csabella?

**Comment by terryjreedy** [comment]
I summarized 'remaining issues' on Sept 8, 2018, and added more on Sept 23 and Dec 14.  I need to check all against current code updated since then.

**Comment by csabella** [comment]
As far as I know, I had made all the requested changes and I had tried to keep up with rebasing.

**Comment by ambv** [comment]
This missed the boat for inclusion in Python 3.9 which accepts security fixes only as of today.
